File: src/main/java/Day1.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

public class Day1 {
    public static void main(String[] args) throws Exception {
        solve();
    }

    private static final Map<String, Integer> NUMBERS_MAP = Map.of(
            "one", 1,
            "two", 2,
            "three", 3,
            "four", 4,
            "five", 5,
            "six", 6,
            "seven", 7,
            "eight", 8,
            "nine", 9
    );

    private static void solve() throws Exception {
        SimpleProfiler profiler = new SimpleProfiler().start();
        try (Stream<String> lines = Files.lines(Paths.get(Day1.class.getResource("day1.txt").toURI()))) {
            int resultPart1 = lines.map(Day1::extractSumPart1).reduce(0, Integer::sum);
            System.out.println(resultPart1);
        }
        profiler.stop();

        profiler.start();
        try (Stream<String> lines = Files.lines(Paths.get(Day1.class.getResource("day1.txt").toURI()))) {
            int resultPart2 = lines.map(Day1::extractSumPart2).reduce(0, Integer::sum);
            System.out.println(resultPart2);
        }
        profiler.stop();
    }

    private static int extractSumPart1(String line) {
        int leftIndex = 0;
        int rightIndex = line.length() - 1;
        String leftNbr = "";
        String rightNbr = "";
        while (leftIndex <= rightIndex) {
            if (Character.isDigit(line.charAt(leftIndex))) {
                leftNbr = String.valueOf(line.charAt(leftIndex));
            } else {
                leftIndex++;
            }
            if (Character.isDigit(line.charAt(rightIndex))) {
                rightNbr = String.valueOf(line.charAt(rightIndex));
            } else {
                rightIndex--;
            }
            if (!leftNbr.isEmpty() && !rightNbr.isEmpty()) {
                break;
            }
        }
        return Integer.parseInt(leftNbr + rightNbr);
    }

    private static int extractSumPart2(String line) {
        List<String> identifiedNbrs = extractAllWords(line);
        String firstNbr = identifiedNbrs.get(0);
        String firstNbrModified = NUMBERS_MAP.containsKey(firstNbr) ? String.valueOf(NUMBERS_MAP.get(firstNbr)) : firstNbr;
        String lastNbr = identifiedNbrs.get(identifiedNbrs.size() - 1);
        String lastNbrModified = NUMBERS_MAP.containsKey(lastNbr) ? String.valueOf(NUMBERS_MAP.get(lastNbr)) : lastNbr;

        return Integer.parseInt(firstNbrModified + lastNbrModified);
    }

    private static List<String> extractAllWords(String line) {
        List<String> wordsInOrder = new ArrayList<>();
        for (int i = 0; i < line.length(); i++) {
            if (Character.isDigit(line.charAt(i))) {
                wordsInOrder.add(String.valueOf(line.charAt(i)));
            } else {
                for (int j = line.length(); j > i; j--) {
                    String subStr = line.substring(i, j);
                    if (NUMBERS_MAP.containsKey(subStr)) {
                        wordsInOrder.add(subStr);
                    }
                }
            }
        }
        return wordsInOrder;
    }
}

File: 2023/python/day1.py
file1 = open('./../src/main/resources/day1.txt', 'r')
lines = file1.readlines()


def part1():
    result = []
    for line in lines:
        result.append(extract_numbers(line))
    print(sum(result))


def extract_numbers(line):
    left_index = 0
    right_index = len(line) - 1
    first_nbr = None
    second_nbr = None
    while left_index <= right_index:
        if line[left_index].isdigit():
            first_nbr = line[left_index]
        else:
            left_index += 1
        if line[right_index].isdigit():
            second_nbr = line[right_index]
        else:
            right_index -= 1
        if first_nbr and second_nbr:
            break
    return int(first_nbr + second_nbr)


def replace_number_words_in_line(line):
    word_to_number_map = {
        "one": "o1e",
        "two": "t2o",
        "three": "th3ee",
        "four": "fo4r",
        "five": "fi5e",
        "six": "s6x",
        "seven": "se7en",
        "eight": "ei8ht",
        "nine": "ni9e"
    }
    for word in word_to_number_map:
        line = line.replace(word, word_to_number_map.get(word))

    return line


def part2():
    result = []
    for line in lines:
        line_replaced = replace_number_words_in_line(line)
        result.append(extract_numbers(line_replaced))
    print(sum(result))


part1()
part2()
